#Tratamiento de variables categóricas

In [0]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

In [0]:
df = pd.read_csv("https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/ecom-expense/Ecom%20Expense.csv")

In [0]:
df.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  
0  Female    Tier 1  4198.385084  
1  Female    Tier 2  4134.976648  
2    Male    Tier 2  5166.614455  
3  Female    Tier 1  7784.447676  
4  Female    Tier 2  3254.160485

In [0]:
dummy_gender = pd.get_dummies(df["Gender"], prefix = "Gender")
dummy_city_tier = pd.get_dummies(df["City Tier"], prefix = "City")

In [0]:
dummy_gender.head()

Gender_Female  Gender_Male
0              1            0
1              1            0
2              0            1
3              1            0
4              1            0

In [0]:
dummy_city_tier.head()

City_Tier 1  City_Tier 2  City_Tier 3
0            1            0            0
1            0            1            0
2            0            1            0
3            1            0            0
4            0            1            0

In [0]:
column_names = df.columns.values.tolist()
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend']

In [0]:
df_new = df[column_names].join(dummy_gender)
column_names = df_new.columns.values.tolist()
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  
0  Female    Tier 1  4198.385084              1            0  
1  Female    Tier 2  4134.976648              1            0  
2    Male    Tier 2  5166.614455              0            1  
3  Female    Tier 1  7784.447676              1            0  
4  Female    Tier 2  3254.160485              1            0

In [0]:
df_new = df_new[column_names].join(dummy_city_tier)
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3  
0            0            0  
1            1            0  
2            1            0  
3            0            0  
4            1            0

In [0]:
feature_cols = ["Monthly Income", "Transaction Time", "Gender_Female", "Gender_Male", "City_Tier 1", "City_Tier 2", "City_Tier 3", "Record"]
#Se añade "Record" para ver su efecto en el resultado, el cual mejora.
#Se puede agregar "Age" para experimentar.

In [0]:
X = df_new[feature_cols]
Y = df_new["Total Spend"]

In [0]:
lm = LinearRegression()
lm.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [0]:
print(lm.intercept_)
print(lm.coef_)

-79.41713030137544
[ 1.47538980e-01  1.54946125e-01 -1.31025013e+02  1.31025013e+02
  7.67643260e+01  5.51389743e+01 -1.31903300e+02  7.72233446e+02]


In [0]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.14753898049205746),
 ('Transaction Time', 0.15494612549589704),
 ('Gender_Female', -131.02501325554584),
 ('Gender_Male', 131.02501325554567),
 ('City_Tier 1', 76.76432601049501),
 ('City_Tier 2', 55.13897430923278),
 ('City_Tier 3', -131.90330031972783),
 ('Record', 772.233445744565)]

In [0]:
lm.score(X,Y)

0.9179923586131016

El modelo puede ser escrito como: <br>
Total_Spend = -79.41713030137544 + 'Monthly Income' * 0.14753898049205746 + 'Transaction Time' * 0.15494612549589704 + 'Gender Female' * -131.02501325554584 + 'Gender Male' * 131.02501325554567 + 'City_Tier 1' * 76.76432601049501 + 'City_Tier 2' * 55.13897430923278 + 'City_Tier 3' * -131.90330031972783 + 'Record' * 772.233445744565
* Si es hombre y vive en CT1: Total_Spend = 128.37220896466724 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545 + 'Record'* 772.2334457445648
* Si es hombre y vive en CT2: Total_Spend = 106.74685726340445 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545 + 'Record'* 772.2334457445648
* Si es hombre y vive en CT3: Total_Spend = -80.29541736555583 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545+'Record'* 772.2334457445648
* Si es mujer y vive en CT1: Total_Spend = -79.41713030137362 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545 - 131.0250132555456+ 76.76432601049527 +'Record'* 772.2334457445648
* Si es mujer y vive en CT2: Total_Spend = -79.41713030137362 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545 - 131.0250132555456+ 55.138974309232474 +'Record'* 772.2334457445648
* Si es mujer y vive en CT3: Total_Spend = -79.41713030137362 + 'Monthly Income' 0.14753898049205738 + 'Transaction Time' 0.15494612549589545 - 131.0250132555456-131.9033003197278 +'Record'* 772.2334457445648

<br> (Se descartan las variables de Gender porque ya sabemos quién es la persona, se cambia el factor del City Tier en el que vive por el que le corresponde, se remueven aquellos donde no vive)

In [0]:
df_new["prediction"] = -79.41713030137362 + df_new['Monthly Income']*0.14753898049205738 + df_new['Transaction Time']* 0.15494612549589545+ df_new['Gender_Female'] * (-131.02501325554567) + df_new['Gender_Male'] * 131.0250132555456+ df_new['City_Tier 1']* 76.76432601049527 +  df_new['City_Tier 2']* 55.138974309232474 + df_new['City_Tier 3']* (-131.9033003197278)+ df_new['Record']* 772.2334457445648

In [0]:
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3   prediction  
0            0            0  4903.696720  
1            1            0  4799.434826  
2            1            0  5157.082504  
3            0            0  8068.012996  
4            1            0  3581.980335

In [0]:
SSD = np.sum((df_new["prediction"] - df_new["Total Spend"])**2)
SSD

1517733985.3408163

In [0]:
RSE = np.sqrt(SSD/(len(df_new)-len(feature_cols)-1))
RSE

803.1318809818165

In [0]:
sales_mean = np.mean(df_new["Total Spend"])
sales_mean

6163.176415976714

In [0]:
error = RSE/sales_mean

In [0]:
error*100

13.031135680294161

In [0]:
#Otra manera de hacer las predicciones de manera no manual
df_new["prediction"] = lm.predict(pd.DataFrame(df_new[feature_cols]))

## Eliminar las variables dummy redundantes
(n-1)

In [0]:
dummy_gender = pd.get_dummies(df["Gender"], prefix="Gender").iloc[:,1:]
dummy_gender.head()

Gender_Male
0            0
1            0
2            1
3            0
4            0

In [0]:
dummy_city_tier = pd.get_dummies(df["City Tier"], prefix="City").iloc[:,1:]
dummy_city_tier.head()

City_Tier 2  City_Tier 3
0            0            0
1            1            0
2            1            0
3            0            0
4            1            0

In [0]:
column_names = df.columns.values.tolist()
df_new = df[column_names].join(dummy_gender)
column_names = df_new.columns.values.tolist()
df_new = df_new[column_names].join(dummy_city_tier)
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Male  City_Tier 2  City_Tier 3  
0  Female    Tier 1  4198.385084            0            0            0  
1  Female    Tier 2  4134.976648            0            1            0  
2    Male    Tier 2  5166.614455            1            1            0  
3  Female    Tier 1  7784.447676            0            0            0  
4  Female    Tier 2  3254.160485            0            1            0

In [0]:
feature_cols = ["Monthly Income", "Transaction Time", "Gender_Male", "City_Tier 2", "City_Tier 3", "Record"]
X = df_new[feature_cols]
Y = df_new["Total Spend"]
lm = LinearRegression()
lm.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [0]:
print(lm.intercept_)

-133.67781754642238


In [0]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.1475389804920575),
 ('Transaction Time', 0.1549461254958966),
 ('Gender_Male', 262.0500265110957),
 ('City_Tier 2', -21.625351701263007),
 ('City_Tier 3', -208.6676263302229),
 ('Record', 772.2334457445636)]

In [0]:
lm.score(X,Y)

0.9179923586131016

Coeficientes con todas las variables en el modelo

* ('Monthly Income', 0.14753898049205738),
* ('Transaction Time', 0.15494612549589545),
* ('Gender_Female', -131.02501325554567),
* ('Gender_Male', 131.0250132555456),
* ('City_Tier 1', 76.76432601049527),
* ('City_Tier 2', 55.138974309232474),
* ('City_Tier 3', -131.9033003197278),
* ('Record', 772.2334457445648)

Coeficientes tras enmascarar las variables dummy pertinentes

* ('Monthly Income', 0.14753898049205744),
* ('Transaction Time', 0.15494612549589631),
* ('Gender_Male', 262.05002651109595),
* ('City_Tier 2', -21.62535170126296),
* ('City_Tier 3', -208.66762633022324),
* ('Record', 772.2334457445635)]

<br>Los cambios se reflejan en

* Gender_Male:
  * antes -> 131.02,
  * después -> 262.05 = ( 131.02 - (-131.02))
* Gender_Female:
  * antes -> -131.02,
  * después -> 0
* CT1:
  * antes -> 76.76,
  * después -> 0
* CT2:
  * antes -> 55.13,
  * después -> -21.62 = (55.13 - 76.76)
* CT3:
  * antes -> -131.90,
  * después -> -208.66 = (-131.90 - 76.76)